### DISCUSSION TITLE
Integration into QuickFolders for French users

### DISCUSSION THREAD

**Discussion by RealRaven2000** [wshanks/tbkeys#189] [Q&A]
I would like to be able to remap Shortcuts for my French users, they use Ctrl+Number (without Shift) to activate one of 10 QuickFolders Tabs:

![image](https://github.com/user-attachments/assets/ec469162-696f-4a5d-ab14-a0e4778fde6f)

The problem is that it is not possible to type a number without SHIFT on [AZERTY keyboards](https://en.wikipedia.org/wiki/AZERTY).
they only allow `CTRL` + `&` `é` `"` `'` `(` `-` `è` `_` `ç` `à` `)` `=` - numbers require SHIFT.

I see that we can use internal Thunderbird commands in a shortcut, would it be possible to access global name spaced functions in experimental  Add-ons as well? I have a global namespace "window.QuickFolders" that could be used. Alternatively, can the shortcuts be remapped to be other shortcuts? 

E.g.:

```json
{
    "ctrl+&": "ctrl+1",
    "ctrl+é": "ctrl+2",
    "ctrl+\"": "ctrl+3",
    "ctrl+'": "ctrl+4",
    "ctrl+(": "ctrl+5",
    "ctrl+-": "ctrl+6",
    "ctrl+è": "ctrl+7",
    "ctrl+_": "ctrl+8",
    "ctrl+ç": "ctrl+9",
    "ctrl+à": "ctrl+0",
}

```

**Comment by RealRaven2000** [comment]
For testing. I have tried the following:

```json
{
    "ctrl+0": "unset",
    "ctrl+1": "memsg:quickfolders@curious.be:{cmd:'shortcut',code:1,isCtrl:true}",
    "ctrl+2": "memsg:quickfolders@curious.be:{cmd:'shortcut',code:2,isCtrl:true}",
}
```

In my (QuickFolders) background page I have added a switch statement for this one:

```js
  messenger.runtime.onMessage.addListener((data, sender) => {
    if (msg_commands.includes(data.command)) {
      return notificationHandler(data, sender); // the result of this is a Promise
    }
    switch (data.command) {
      case "shortcut":
        console.log("QuickFolders: Received shortcut:", { data , sender });
        break;
    }
  });
```

obviously I am trying a json object as the "message". not sure how a "string message" would work syntactially? An example would be great!


I also tried:
`"ctrl+2": "memsg:quickfolders@curious.be:'TEST'"`
and
`"ctrl+2": "memsg:quickfolders@curious.be:TEST"`

but `messenger.runtime.onMessage` in my Add-on _never receives anything_. What am I doing wrong? The Add-on id is from my manifest:

```json
  "manifest_version" : 2,
  "applications" : {
    "gecko" : {
      "id" : "quickfolders@curious.be",
      "strict_min_version" : "111.0",
      "strict_max_version" : "132.*"
    }
  },
```

**Comment by morat523035** [comment]
Here is how to execute the "Toggle Folder tree visibility" command in the manifest.json file using tbkeys. (useful for single letter shortcut without alt or ctrl)

```
{
  "v": "(function () { var extension = window.ExtensionParent.GlobalManager.getExtension('quickfolders@curious.be'); extension.shortcuts.onCommand('toggle-foldertree'); })();"
}
```

QuickFolders
http://addons.thunderbird.net/thunderbird/addon/3254

Reference (see toggle-foldertree and commands.toggleFolderTree)
http://github.com/RealRaven2000/QuickFolders/blob/ESR115/manifest.json
http://github.com/RealRaven2000/QuickFolders/blob/ESR115/qf-background.js
http://github.com/RealRaven2000/QuickFolders/blob/ESR115/_locales/en/messages.json

The extension id is listed in the Debugging tab.

i.e. menu bar > tools > developer tools > debug addons

Here is how to view the command names with descriptions for an addon using the error console. The code snippet could be useful for a tbkeys user that wants to execute an addon command, but doesn't know the command name.

i.e. menu bar > tools > developer tools > error console

```
(function () {
  var extension = window.ExtensionParent.GlobalManager.getExtension('quickfolders@curious.be');
  extension.shortcuts.manifestCommands.forEach(function (command, name) {
    console.log(name + ": " + command.description);
  });
})();
```

Output for QuickFolders:

`toggle-foldertree: Toggle Folder tree visibility`
`focus-foldersbox: Keyboard focus on QuickFolders toolbar`

Similar issue
http://github.com/wshanks/tbkeys/issues/211
